qcog_python_client/qcog/pytorch/handler.py

```diff
@@ -0,0 +1,131 @@
+"""Defines the interface for the chain of responsibility pattern.
+
+Commands get dispatched through the chain of handlers. Each handler
+checks if it can handle the command. If it can, it executes the command.
+If the handler returns a new command, it gets dispatched again though the chain.
+
+Each handler implements an automatic retry mechanism.
+The number of attempts and the time to wait can be configured for each handler.
+
+In case of error, the `revert` method is called to undo any changes made by the handler.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import enum
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Coroutine, Generic, TypeAlias, TypeVar
+
+from pydantic import BaseModel
+
+
+class BoundedCommand(BaseModel):
+    """Command type."""
+
+    command: Command
+    dispatch_next: bool = False
+
+
+CommandPayloadType = TypeVar("CommandPayloadType", bound=BoundedCommand)
+ToolName: TypeAlias = str
+ToolFn: TypeAlias = Callable[..., Coroutine[Any, Any, dict]]
+
+
+class Command(enum.Enum):
+    """Command enum."""
+
+    save_parameters = "save_parameters"
+    discover = "discover"
+    validate = "validate"
+    upload = "upload"
+    revert_all = "revert_all"
+    train = "train"
+
+
+class Handler(ABC, Generic[CommandPayloadType]):
+    """Interface for the chain of responsibility pattern."""
```

Comment:
So I guess I have a question about the data structure, and we should be clear. It looks like on the surface to be a linked list where every command has a pointer to the next command. However it also all has a pointer to the head of the chain rather than the previous command. Why are we doing this rather than implementing a linked list pattern?

Reply:
So chain of responsibility usually goes in one direction
client - Command -> [ Handler[Head] -> Handler[1] -> Handler[2] ]
the client issue the command but doesn't know which handler will perform the command.
To be more flexible I added the reference of the head to each handler, so the handler can issue a new command for the chain. If a command is returned from an handler and `dispatch_next` is explicitly `True`, it will be dispatched from the head of the list. 

https://github.com/qognitive/qcog-python-client/blob/2fe470fdadcd92e57fa84d5b5d51224eaff06b20/qcog_python_client/qcog/pytorch/handler.py?plain=0#L118-L131